feat(#20): Robot grabber (#47)

42piotrnycz opened the pull request:
closes #20

42piotrnycz committed:
oops

42piotrnycz committed:
added-route-movement-grabbing-releasing

Automatic Grab Mode - releasing doesn't work (probably because in the moment of releasing the items get automatically picked up as they enter the GrabberSphere)

Moving on slopes isn't implemented

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+            {
+                grabbedItem.transform.position = transform.position + grabOffset;
+            }
+        }
```
`if (grabbedItem is not null)` use this as it is more explicit

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+            lastItem = grabbedItem;
+            if (grabbedItem != null)
+            {
+                Vector3 pos = grabbedItem.transform.position;
```
I would reverse the if statement to `if (grabbedItem is null) return;` this will get rid of nesting and would make this function more readable.

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+                LayerMask playerMask = 1 << LayerMask.NameToLayer("Enemy");
+
+                if (Physics.Raycast(ray, out hit, Mathf.Infinity, ~playerMask))
+                {
```
We can get rid of `RaycastHit hit;` if we define it inside the if statement like this `if (Physics.Raycast(ray, out RaycastHit hit...` or  `if (Physics.Raycast(ray, out var hit...` both works for me.

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+            foreach (Collider nearItem in nearItems) 
+            {
+                if (nearItem.gameObject.tag == "Item")
+                {
```
I would use LINQ here instead of if like `foreach(Collider nearItem in nearItems.Where(item => item.gameObject.tag == "Item"))`. Then we can get rid of the `if (nearItem.gameObject.tag == "Item")`. I'm not sure about the performance tho. @PatrykPaluch Thoughts?

PatrykPaluch replied:
I think in linq you only pay for a function call (lambda), so there should be no performance difference, but I prefer `foreach(..){ if(...) }` approach. You can always negate `if` condition: `if(nearItem.gameObject.tag != "Item") continue;`. And in Unity you can use `nearItem.CompareTag("Item")` instead. Its shorter and faster.

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            {
+                UpdateRoute();
+                if (grabber != null) GrabOrRelease();
+            }
```
Use `grabber is not null`

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            Vector3 velocity = directionToPoint * robotSpeed;
+            rb.velocity = velocity;
+        }
+        private void ProcessRotation()
```
Enter after the `}` to separate the functions

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            transform.rotation = rotation;
+        
+        }
+        private bool IsGrounded()
```
Enter here too

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            int add_amount = routeForward ? 1 : -1;
+            routeIndex += add_amount;
+        }
+        private void GrabOrRelease()
```
Enter here too

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+                grabber.GrabItem();
+            }
+            
+        }
+        private bool CheckIfArrived(Vector3 point)
```
Enter here too

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+                directionToPoint += Vector3.down * gravity * Time.fixedDeltaTime;
+            }
+            directionToPoint.Normalize();
+
```
Same as above. The directionToPoint is already normalized. Move this inside the `else` block so we only invoke this function when needed

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            }
+            
+            Vector3 directionToPoint = GetDirectionToPoint(transform.position, route[routeIndex].waypoint);
+            directionToPoint.y = 0f;
```
Doesn't GetDirectionToPoint return the Vec3 with Y = 0? Do we need to repeat this logic here?

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+        private void UpdateRoute()
+        {
+            if (routeForward == true)
+            {
```
I would rename the `routeForward` function to something more... 'boolish'

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            else if (route[routeIndex].grab && (grabber.grabbedItem == null))
+            {
+                grabber.GrabItem();
+            }
```
We don't need the first `else if`. Just make this `if(...) return;` and then `if (route[rou...)`

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+        private void GrabOrRelease()
+        {
+            if (grabber.automaticGrabMode) return;
+            else if (route[routeIndex].release && (grabber.grabbedItem != null))
```
`grabber.grabbedItem is not null` and remove the needless `()`

Vie0n commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+            {
+                grabber.ReleaseItem();
+            }
+            else if (route[routeIndex].grab && (grabber.grabbedItem == null))
```
`grabber.grabbedItem is null` and remove the needless `()`

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+        public float releaseDistance;
+
+        private RouteMovement routeMovement;
+        public void Start()
```
Enter here after the `private RouteMovement routeMovement;`

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+                    GrabItem();
+                }
+            }
+        }
```
I would split the ifs into separate conditions.
- First `if (!automaticGrabMode) return;`
- Then the sec `if (grabbedItem is not null) return;`
- And the last `if (lastItem is not collider.gameObject && collider.CompareTag("Item")) GrabItem();`
This will make the code easier to read and would remove the needless `()` parentheses.

Vie0n commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+
+        public void GrabItem()
+        {   
+            if (grabbedItem == null)
```
`if (grabbedItem is null)...`

PatrykPaluch commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+                RaycastHit hit;
+                LayerMask playerMask = 1 << LayerMask.NameToLayer("Enemy");
+
+                if (Physics.Raycast(ray, out hit, Mathf.Infinity, ~playerMask))
```
Avoid accessing Layers through names or indexes. Create a class field of type `LayerMask`: `[SerializeField] private LayerMask enemyLayer;` and set this layer in the inspector. It's safer, coz if someone changes the layer name, then the inspector will update accordingly.

Also, you named this variable `playerMask` but you assigned "enemy" layer to it. Maybe it would be better if you remove the negation (`~`) from the Raycast function argument and you just assign appropriate layers from the inspector.

I'm not sure if this layer variable name is correct, I'm uncertain about its intended meaning

42piotrnycz replied:
Changed the name to 'releaseExcludedLayer' so it would depict what's it's exactly used for (excluding from raycast).

PatrykPaluch commented on Assets/Scripts/Robots/Grabber.cs:
```diff
@@ -0,0 +1,114 @@
+            float nearestDistance = float.MaxValue;
+            GameObject nearestItem = null;
+
+            Collider[] nearItems = Physics.OverlapSphere(transform.position, grabDistance);
```
The `OverlapSphere` will allocate memory for array, every time it is called. You use this function inside `OnTrigggerEnter` and this script is intended to be placed on moving object, so there will be a lot of `OnTriggerEnter` events. That means `OverlapSpahre` will allocate a lot of memory. We are not in stage where we need to worry about performance, but there is simple solution to that problem so we can try to address this issue right in this PR. You can simply change `OverlapSphere` to `OverlapSphereNonAlloc`, it will be something like this:

```cs
private Collider[] nerestItemsBuffer = new Collider[5];
```
```cs
int itemsCount = Physics.OverlapSphereNonAlloc(transform.position, grabDistance, nerestItemsBuffer);

for (int i = 0 ; i < itemsCount ; i++) {
 //...
}
```
Note that now you can't use `nerestItemsBuffer.Length` because it will be size of the buffer instead of count of overlaped items. Now, you must use `for` loop instead of `foreach`. One more disadvantage is that you must specify max items that can be detected by single `OverlapSphereNonAlloc` call. In most situations, it's not an issue.

Beside that, consider specifying `LayerMash` for the OverlapSphare (like comment before - declare `LayerMask` as private class field).

[Physics.OverlapSphereNonAlloc](https://docs.unity3d.com/ScriptReference/Physics.OverlapSphereNonAlloc.html)

42piotrnycz replied:
Will implement that in the future, but right now (as we lack time) stayed with the previous solution.

PatrykPaluch commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+        public int routeIndex = 0;
+        public Route[] route;
+        private bool routeForward = true;
+        #endregion route
```
You can use `[Header("route")]` to display this nicely in inspector ;)

```cs
[Header("Route")]
public bool loopMode;
public int routeIndex = 0;
public Route[] route;
```

PatrykPaluch commented on Assets/Scripts/Robots/RouteMovement.cs:
```diff
@@ -0,0 +1,135 @@
+        {
+            rb = GetComponent<Rigidbody>();
+            grabber = GetComponentInChildren<Grabber>();
+        }
```
add `[RequireComponent(typeof(Rigidbody), typeof(Grabber))]` attribute to class or add null check here

PatrykPaluch commented on ProjectSettings/TagManager.asset:
```diff
@@ -7,11 +7,12 @@ TagManager:
   - TransparentFX
   - Ignore Raycast
-  - 
+  - Floor
```
I'm not sure about that. What if we make a smooth surface (eg. sphere) that can act like wall and floor (and ceiling) at the same time?

42piotrnycz replied:
Deleted the layer as it was not used at all.

42piotrnycz committed:
after-code-review

still haven't fixed all of the issues

42piotrnycz committed:
addressed-github-issues

fixed slope movement
fixed bug where the robot released the item, and the immediately picked it up
addressed issues from github

42piotrnycz committed:
addressed-github-issues-2
